File: windows-app/Plantoir.Core/Scripting/TranscriptBuilder.cs
using System.Text;

namespace Plantoir.Core.Scripting;

/// <summary>
/// Turns raw terminal output into clean display lines.
///
/// A pseudo console converts every "\n" the script prints into "\r\n", so
/// CR followed by LF is an ordinary line ending. Only a LONE carriage
/// return is a spinner redrawing its line — it restarts the current line.
/// ANSI escape sequences (colours, cursor movement, OSC titles) are
/// stripped before the line machine runs.
/// </summary>
public sealed class TranscriptBuilder
{
    public const int MaximumRetainedLines = 4000;

    private readonly List<string> _lines = new();
    private readonly StringBuilder _currentLine = new();
    private bool _hasPendingCarriageReturn;
    private string? _cachedDisplayText;
    private long _version;

    /// <summary>Completed lines, oldest first (capped at MaximumRetainedLines).</summary>
    public IReadOnlyList<string> Lines => _lines;

    /// <summary>The unterminated line under construction.</summary>
    public string CurrentLine => VisibleCurrentLine;

    /// <summary>
    /// The line under construction, unless it is machinery.
    ///
    /// <para>A <c>PLANTOIR_HEALTH:</c> line is a JSON blob, and a teacher
    /// reads this console (CLAUDE.md rule 1). Nothing is lost by hiding it:
    /// <c>scripts/site_health.py</c> prints the human sentence separately,
    /// beside the marker, and that sentence is left alone.</para>
    ///
    /// <para>Hidden here rather than only at <see cref="PushLine"/> because a
    /// pseudo console hands over whatever bytes are ready — the marker and its
    /// payload sit in the current line for as long as it takes the rest of the
    /// line to arrive, and the console renders it in the meantime.</para>
    /// </summary>
    private string VisibleCurrentLine
    {
        get
        {
            string line = _currentLine.ToString();
            return CarriesTheHealthMarker(line) ? "" : line;
        }
    }

    /// <summary>
    /// Whether this line is a health finding's machine-readable half.
    ///
    /// <para>The whole line goes, not just the marker onward: the launchers can
    /// glue the marker to the tail of their own chatter, and what precedes it
    /// there is progress noise rather than a sentence a teacher needs. The
    /// findings themselves are never at risk — <c>ScriptRunner.ReceiveOutput</c>
    /// hands the RAW text to <c>CollectHealthFindings</c>, and only then to
    /// this builder.</para>
    /// </summary>
    private static bool CarriesTheHealthMarker(string line) =>
        line.Contains(Plantoir.Core.Models.SiteHealthFinding.Marker, StringComparison.Ordinal);

    /// <summary>Monotonic counter bumped on every append — cheap change detection.</summary>
    public long Version => _version;

    public void Append(string rawText)
    {
        _cachedDisplayText = null;
        _version++;
        string cleaned = StripControlSequences(rawText);
        foreach (char c in cleaned)
        {
            if (_hasPendingCarriageReturn)
            {
                _hasPendingCarriageReturn = false;
                if (c == '\n') { PushLine(); continue; }
                _currentLine.Clear();               // lone CR: spinner redraw
            }
            if (c == '\n') { PushLine(); }
            else if (c == '\r') { _hasPendingCarriageReturn = true; }
            else { _currentLine.Append(c); }
        }
    }

    private void PushLine()
    {
        string line = _currentLine.ToString();
        _currentLine.Clear();
        // Dropped rather than retained-and-filtered-on-read, so it cannot reach
        // the problem report either: SaveRunTranscript copies Lines wholesale
        // into the report store, and a payload a teacher may not see is not one
        // support should be handed. The trail keeps the check's NAME, which is
        // the part anybody reading the report back would search for.
        if (CarriesTheHealthMarker(line)) return;
        _lines.Add(line);
        if (_lines.Count > MaximumRetainedLines)
            _lines.RemoveRange(0, _lines.Count - MaximumRetainedLines);
    }

    public string DisplayText
    {
        get
        {
            if (_cachedDisplayText is null)
            {
                string current = VisibleCurrentLine;
                var all = current.Length > 0
                    ? _lines.Append(current)
                    : _lines;
                _cachedDisplayText = string.Join("\n", all);
            }
            return _cachedDisplayText;
        }
    }

    /// <summary>The tail of the transcript, built without joining everything.</summary>
    public string RecentText(int maximumCharacters)
    {
        var collected = new List<string>();
        int budget = maximumCharacters;
        string current = VisibleCurrentLine;
        if (current.Length > 0) { collected.Add(current); budget -= current.Length + 1; }
        for (int i = _lines.Count - 1; i >= 0 && budget > 0; i--)
        {
            collected.Add(_lines[i]);
            budget -= _lines[i].Length + 1;
        }
        collected.Reverse();
        return string.Join("\n", collected);
    }

    /// <summary>
    /// Strips ANSI CSI (ESC [ ... final letter), OSC (ESC ] ... BEL), other
    /// two-byte escapes, and control characters below 0x20 except \n \r \t.
    /// </summary>
    public static string StripControlSequences(string text)
    {
        var result = new StringBuilder(text.Length);
        int i = 0;
        while (i < text.Length)
        {
            char c = text[i];
            if (c == '\x1b')
            {
                if (i + 1 < text.Length && text[i + 1] == '[')
                {
                    int j = i + 2;
                    while (j < text.Length && !char.IsAsciiLetter(text[j])) j++;
                    i = j + 1;
                    continue;
                }
                if (i + 1 < text.Length && text[i + 1] == ']')
                {
                    int j = i + 2;
                    while (j < text.Length && text[j] != '\x07') j++;
                    i = j + 1;
                    continue;
                }
                i += 2;
                continue;
            }
            if (c < ' ' && c != '\n' && c != '\r' && c != '\t') { i++; continue; }
            result.Append(c);
            i++;
        }
        return result.ToString();
    }
}

File: mac-app/QuartzTeachers/Scripting/TranscriptBuilder.swift
import Foundation

/// Turns raw terminal output into clean display lines.
///
/// The scripts produce terminal control sequences (colours, spinners that
/// redraw the same line with carriage returns, cursor movement). Teachers
/// should see tidy text, so this type strips ANSI escape sequences and
/// treats a carriage return as "start this line over", which collapses
/// spinner animation into its final state.
/// Holds the joined transcript text so reading it mutates nothing.
final class TranscriptTextCache {

    // MARK: - Stored properties

    var text: String?
}

struct TranscriptBuilder {

    // MARK: - Stored properties

    /// Completed lines of output, oldest first.
    var lines: [String] = []

    /// The line currently being assembled (not yet ended with a newline).
    var currentLine: String = ""

    /// True when the previous chunk ended with a carriage return whose
    /// meaning (line ending vs. spinner redraw) depends on what follows.
    var hasPendingCarriageReturn: Bool = false

    /// The most recent lines are all a teacher ever reads, and keeping
    /// every line of a long publish would grow without bound.
    static let maximumRetainedLines: Int = 4000

    /// Cached joined text: `displayText` is read several times per
    /// refresh, and rebuilding it each time is what stalls the UI.
    ///
    /// The cache lives in a REFERENCE box so that reading `displayText`
    /// mutates nothing. This transcript is stored on an observed object
    /// and read from view bodies: a mutating getter would write to
    /// observed state during view evaluation, invalidating the view,
    /// which re-reads, which writes again — an endless loop that freezes
    /// the interface. `append` swaps in a fresh box, so copies of this
    /// struct never share a stale cache.
    private var cache: TranscriptTextCache = TranscriptTextCache()

    // MARK: - Computed properties

    /// The full transcript as one display string (cached).
    var displayText: String {
        if let text = cache.text {
            return text
        }
        var allLines: [String] = lines
        if !currentLine.isEmpty {
            allLines.append(currentLine)
        }
        let text: String = allLines.joined(separator: "\n")
        cache.text = text
        return text
    }

    // MARK: - Functions

    /// The tail of the transcript, without joining every line.
    func recentText(maximumCharacters: Int) -> String {
        var collected: [String] = []
        var characterCount: Int = 0
        if !currentLine.isEmpty {
            collected.append(currentLine)
            characterCount += currentLine.count
        }
        var index: Int = lines.count - 1
        while index >= 0 && characterCount < maximumCharacters {
            let line: String = lines[index]
            collected.append(line)
            characterCount += line.count + 1
            index -= 1
        }
        return collected.reversed().joined(separator: "\n")
    }

    /// Feed a chunk of raw output from the PTY into the transcript.
    ///
    /// A pseudo-terminal turns every "\n" the script prints into "\r\n",
    /// so a carriage return followed by a newline is an ordinary line
    /// ending. Only a LONE carriage return is a spinner redrawing its
    /// line, which is when the current line restarts.
    mutating func append(rawText: String) {
        // A fresh box per mutation: copies of this struct must never
        // share a cache that one of them later fills in.
        cache = TranscriptTextCache()
        let cleaned: String = TranscriptBuilder.strippingControlSequences(from: rawText)
        // Work scalar-by-scalar: Swift groups "\r\n" into a SINGLE
        // Character (grapheme cluster), which would hide line endings.
        let newlineScalar: Unicode.Scalar = "\n"
        let carriageReturnScalar: Unicode.Scalar = "\r"
        for scalar in cleaned.unicodeScalars {
            if hasPendingCarriageReturn {
                hasPendingCarriageReturn = false
                if scalar == newlineScalar {
                    // "\r\n": a normal line ending — and THE one real output
                    // takes, because this comes from a PTY. The health-line
                    // filter has to be here as well as in the plain "\n"
                    // branch below; it was only below, so every test passed
                    // (they all supplied "\n") while a real build showed the
                    // teacher the raw JSON.
                    appendUnlessMachineReadable(currentLine)
                    currentLine = ""
                    continue
                }
                // A lone "\r": spinner redraw — restart the line, then
                // process the current scalar normally below.
                currentLine = ""
            }
            if scalar == newlineScalar {
                appendUnlessMachineReadable(currentLine)
                currentLine = ""
                if lines.count > TranscriptBuilder.maximumRetainedLines {
                    lines.removeFirst(lines.count - TranscriptBuilder.maximumRetainedLines)
                }
            } else if scalar == carriageReturnScalar {
                hasPendingCarriageReturn = true
            } else {
                currentLine.unicodeScalars.append(scalar)
            }
        }
    }

    /// Adds a finished line, unless it is one of the machine-readable ones.
    ///
    /// Rule 1: the interface never names the machinery, and a raw JSON blob is
    /// machinery. Nothing is lost — the toolchain prints the teacher-facing
    /// sentence separately, and `ScriptRunner.receiveOutput` reads the raw text
    /// for findings BEFORE handing it here, precisely so this can drop them.
    ///
    /// One function called from BOTH line endings, because having the check in
    /// only one of them is exactly the bug this replaced.
    private mutating func appendUnlessMachineReadable(_ line: String) {
        if SiteHealthFinding.isMarkerLine(line) {
            return
        }
        lines.append(line)
    }

    /// Removes ANSI escape sequences and stray control characters,
    /// keeping newlines, carriage returns, and tabs.
    ///
    /// Operates on Unicode scalars, not Characters, because Swift folds
    /// "\r\n" into one Character and would misclassify it as a control
    /// character to remove.
    static func strippingControlSequences(from text: String) -> String {
        var result: String = ""
        let scalars: [Unicode.Scalar] = Array(text.unicodeScalars)
        let escapeScalar: Unicode.Scalar = "\u{1B}"
        let bellScalar: Unicode.Scalar = "\u{07}"
        var index: Int = 0
        while index < scalars.count {
            let scalar: Unicode.Scalar = scalars[index]
            if scalar == escapeScalar {
                // Escape sequence: skip "ESC [ ... final-letter" (CSI) or
                // "ESC ] ... BEL" (OSC), or a single following scalar.
                let nextIndex: Int = index + 1
                if nextIndex < scalars.count && scalars[nextIndex] == "[" {
                    var scanIndex: Int = nextIndex + 1
                    while scanIndex < scalars.count {
                        let scanned: Unicode.Scalar = scalars[scanIndex]
                        let isFinalLetter: Bool = (scanned >= "A" && scanned <= "Z") || (scanned >= "a" && scanned <= "z")
                        if isFinalLetter {
                            break
                        }
                        scanIndex += 1
                    }
                    index = scanIndex + 1
                    continue
                }
                if nextIndex < scalars.count && scalars[nextIndex] == "]" {
                    var scanIndex: Int = nextIndex + 1
                    while scanIndex < scalars.count {
                        let scanned: Unicode.Scalar = scalars[scanIndex]
                        if scanned == bellScalar {
                            break
                        }
                        scanIndex += 1
                    }
                    index = scanIndex + 1
                    continue
                }
                index = index + 2
                continue
            }
            let isControl: Bool = scalar.value < 32
            let isKeeper: Bool = scalar == "\n" || scalar == "\r" || scalar == "\t"
            if isControl && !isKeeper {
                index += 1
                continue
            }
            result.unicodeScalars.append(scalar)
            index += 1
        }
        return result
    }
}
